In [1]:
!pip install nltk

In [2]:
#
# Import csv data from script: Create bigrams with sentence index number and POS
#

import csv

forward_bigrams = []

with open('input - Create bigrams with sentence index number and POS.csv', newline='\n') as csvfile:
    
    csv_reader = csv.reader(csvfile, delimiter=',')
    
    for row in csv_reader:

        sentence_index = row[0]
        bigram_as_string = row[1]
        word_pos_tag_one = row[2]
        word_pos_tag_two = row[2]
        
        forward_bigrams.append((sentence_index,bigram_as_string,word_pos_tag_one,word_pos_tag_two))
        
forward_bigrams

[('1', 'electric installation', 'JJ', 'JJ'),
 ('3', 'chapter apply', 'NN', 'NN'),
 ('3', 'vessel construct', 'NNS', 'NNS'),
 ('4', 'vessel construct', 'NNS', 'NNS'),
 ('4', 'rule applicable', 'NNS', 'NNS'),
 ('4', 'unless otherwise', 'IN', 'IN'),
 ('4', 'otherwise provide', 'RB', 'RB'),
 ('4', 'individual provision', 'JJ', 'JJ'),
 ('5', 'substantially alter', 'RB', 'RB'),
 ('5', 'vessel shall', 'NN', 'NN'),
 ('5', 'shall comply', 'MD', 'MD'),
 ('5', 'new vessel', 'JJ', 'JJ'),
 ('8', 'fire safety', 'NNP', 'NNP'),
 ('8', 'vessel construct', 'NNS', 'NNS'),
 ('8', 'structural fire', 'JJ', 'JJ'),
 ('8', 'fire protection', 'NN', 'NN'),
 ('8', 'nordic boat', 'NNP', 'NNP'),
 ('8', 'boat standard', 'NNP', 'NNP'),
 ('8', 'unless otherwise', 'IN', 'IN'),
 ('8', 'otherwise provide', 'RB', 'RB'),
 ('8', 'individual provision', 'JJ', 'JJ'),
 ('9', 'vessel construct', 'NNS', 'NNS'),
 ('9', 'structural fire', 'JJ', 'JJ'),
 ('9', 'fire protection', 'NN', 'NN'),
 ('9', 'rule applicable', 'NNS', 'NNS'),


# Find known words

In [6]:
#
# Hypotesis: We must know what to look for.
# Here is what we know.
#

# What are the things we know?
list_of_known_things = ['vessel', 'emergency light', 'pipe system', 'pipe', 'fire extinguisher']

# What are the data properties we know?
list_of_known_data_properties = ['length', 'overall length', 'closer', 'place closer', 'constructed after', 'constructed before', 'least']

# What are the data types we know?
#list_of_known_data_types = ['cm', 'metre', 'voltage', 'mm']

# The big question: Do we find any of these three variables in the same sentence?
# Let's find out.

# Idea: Look for all the things we know.
# Next, in a selected line,
# if a data type is identified, 
# but at known data property is unknown, 
# then make a human identify the unknown data property.

## Find known things from vocabulary

In [4]:
# Get a bigram list without duplicates

bigram_list_with_duplicates = []

for (sentence_index,bigram_as_string,word_pos_tag_one,word_pos_tag_two) in forward_bigrams:
    bigram_list_with_duplicates.append(bigram_as_string)
    
forward_bigrams_without_duplicates = list(set(bigram_list_with_duplicates))
forward_bigrams_without_duplicates

['space shall',
 'frost shall',
 'fire door',
 'smoke gas',
 'glass shall',
 'inside planking',
 'easily render',
 'centimetre thickness',
 'approved type',
 'regulator shall',
 'require alarm',
 'overflow opening',
 'well ventilate',
 'test cock',
 'gas use',
 'two approve',
 'shall comply',
 'alarm system',
 'shall satisfy',
 'chapter apply',
 'pipe make',
 'insulation shall',
 'qualify material',
 'possible away',
 'connection make',
 'suitable place',
 'installation shall',
 'seawater pipe',
 'stove without',
 'nordic boat',
 'wool twine',
 'fill automatically',
 'release equipment',
 'bilge pipe',
 'bulkhead shall',
 'fuel oil',
 'voltage drop',
 'self-closing type',
 'closing valve',
 'whole floor',
 'approve fire-extinguishing',
 'readily ignite',
 'low edge',
 'maritime electrical',
 'convey oil',
 'vessel shall',
 'use shall',
 'time interval',
 'cabin shall',
 'flammable liquid',
 'possible collection',
 'stove shall',
 'first paragraph',
 'power machinery',
 'release device'

In [9]:
#
# Here is a list over known things
#

[bigram for bigram in forward_bigrams_without_duplicates if bigram in list_of_known_things]

['fire extinguisher', 'emergency light']

In [3]:
for (sentence_index,bigram_as_string,word_pos_tag_one,word_pos_tag_two) in forward_bigrams:
    print(bigram_as_string)

electric installation
chapter apply
vessel construct
vessel construct
rule applicable
unless otherwise
otherwise provide
individual provision
substantially alter
vessel shall
shall comply
new vessel
fire safety
vessel construct
structural fire
fire protection
nordic boat
boat standard
unless otherwise
otherwise provide
individual provision
vessel construct
structural fire
fire protection
rule applicable
gasfired installation
installation shall
shall satisfy
safety precaution
gasfired installation
liquefied hydrocarbon
hydrocarbon gas
gas use
board vessel
oil-fired stove
electric power
power supply
supply shall
oil supply
smoke gas
flame go
failure occur
stove without
without electric
electric power
power supply
supply shall
oil supply
smoke gas
flame go
drip regulation
regulation stove
regulation valve
observation hole
automatic shutoff
oil supply
oil consumption
oil regulation
regulation valve
installation instruction
instruction shall
shall indicate
maximum permit
permit difference
o

In [10]:
# Get identified bigrams with sentence index number

forward_identified_bigrams_with_sentence_index_number = []

for (sentence_index,bigram_as_string,word_pos_tag_one,word_pos_tag_two) in forward_bigrams:
    
    if bigram_as_string in list_of_known_things:
        forward_identified_bigrams_with_sentence_index_number.append((sentence_index,bigram_as_string))
    
forward_identified_bigrams_with_sentence_index_number

[('90', 'fire extinguisher'),
 ('90', 'fire extinguisher'),
 ('91', 'fire extinguisher'),
 ('91', 'fire extinguisher'),
 ('92', 'fire extinguisher'),
 ('93', 'fire extinguisher'),
 ('94', 'fire extinguisher'),
 ('97', 'fire extinguisher'),
 ('137', 'emergency light')]

In [ ]:
# TODO: Write to CSV file

In [ ]:
## Find 

In [266]:
#
# Find known data types in line
#

known_data_types_in_line = []

for word_we_know in list_of_known_data_types:
    
    # result
    # word_we_know -> data type example: cm
    # get sentence index
    index_of_lines_with_a_match = []
    
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:

        if word_in_text == word_we_know:
            index_of_lines_with_a_match.append(line_index_number)
            #print(word_in_text + " " + str(line_index_number))
            
    if index_of_lines_with_a_match:
        known_data_types_in_line.append((word_we_know, index_of_lines_with_a_match))

known_data_types_in_line

[('cm', [38, 42, 44, 45, 48, 49, 49, 52, 89]),
 ('metre', [35, 41, 46, 64, 70, 76, 77, 105, 106, 107, 109, 112]),
 ('voltage', [105, 105, 106, 106, 107, 107, 109, 109, 113]),
 ('mm', [40])]

In [267]:
#
# Find known data properties in line
#

known_data_properties_in_line = []

for word_we_know in list_of_known_data_properties:
    
    index_of_lines_with_a_match = []
    
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:

        if word_in_text == word_we_know:
            index_of_lines_with_a_match.append(line_index_number)
            #print(word_in_text + " " + str(line_index_number))
            
    if index_of_lines_with_a_match:
        known_data_properties_in_line.append((word_we_know, index_of_lines_with_a_match))

known_data_properties_in_line

[('length', [35, 64, 70, 76, 77, 105, 106, 107, 109, 112]),
 ('closer', [48, 52]),
 ('least',
  [38,
   40,
   41,
   42,
   44,
   45,
   46,
   49,
   77,
   77,
   77,
   78,
   78,
   79,
   79,
   83,
   88,
   89,
   89,
   112])]

In [268]:
# List the findings

findings = []

for (word_we_know, index_of_lines_with_a_match) in known_data_types_in_line:
    for index_of_line in index_of_lines_with_a_match:
        findings.append((index_of_line,word_we_know))

sortedList = sorted(findings, key = lambda y: (y[0]))
sortedList

[(35, 'metre'),
 (38, 'cm'),
 (40, 'mm'),
 (41, 'metre'),
 (42, 'cm'),
 (44, 'cm'),
 (45, 'cm'),
 (46, 'metre'),
 (48, 'cm'),
 (49, 'cm'),
 (49, 'cm'),
 (52, 'cm'),
 (64, 'metre'),
 (70, 'metre'),
 (76, 'metre'),
 (77, 'metre'),
 (89, 'cm'),
 (105, 'metre'),
 (105, 'voltage'),
 (105, 'voltage'),
 (106, 'metre'),
 (106, 'voltage'),
 (106, 'voltage'),
 (107, 'metre'),
 (107, 'voltage'),
 (107, 'voltage'),
 (109, 'metre'),
 (109, 'voltage'),
 (109, 'voltage'),
 (112, 'metre'),
 (113, 'voltage')]

In [269]:
# Create dictionary with findings of known_data_types_in_line
# The dictionary key is the index number of the line

from collections import defaultdict

dictionary_with_known_data_types_in_a_sentence = defaultdict(int)

for (word_we_know, index_of_lines_with_a_match) in known_data_types_in_line:
    
    for index_of_line in index_of_lines_with_a_match:
        
        if index_of_line not in dictionary_with_known_data_types_in_a_sentence:
            
            print('First detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add to dictionary
            dictionary_with_known_data_types_in_a_sentence[index_of_line] = [word_we_know]
            
        else: # a detection in sentence is already added
            
            print('New detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add if it is a new word
            if word_we_know not in dictionary_with_known_data_types_in_a_sentence[index_of_line]:
                dictionary_with_known_data_types_in_a_sentence[index_of_line].append(word_we_know)
                
dictionary_with_known_data_types_in_a_sentence

First detection in sentence 38 - cm
First detection in sentence 42 - cm
First detection in sentence 44 - cm
First detection in sentence 45 - cm
First detection in sentence 48 - cm
First detection in sentence 49 - cm
New detection in sentence 49 - cm
First detection in sentence 52 - cm
First detection in sentence 89 - cm
First detection in sentence 35 - metre
First detection in sentence 41 - metre
First detection in sentence 46 - metre
First detection in sentence 64 - metre
First detection in sentence 70 - metre
First detection in sentence 76 - metre
First detection in sentence 77 - metre
First detection in sentence 105 - metre
First detection in sentence 106 - metre
First detection in sentence 107 - metre
First detection in sentence 109 - metre
First detection in sentence 112 - metre
New detection in sentence 105 - voltage
New detection in sentence 105 - voltage
New detection in sentence 106 - voltage
New detection in sentence 106 - voltage
New detection in sentence 107 - voltage
New d

defaultdict(int,
            {38: ['cm'],
             42: ['cm'],
             44: ['cm'],
             45: ['cm'],
             48: ['cm'],
             49: ['cm'],
             52: ['cm'],
             89: ['cm'],
             35: ['metre'],
             41: ['metre'],
             46: ['metre'],
             64: ['metre'],
             70: ['metre'],
             76: ['metre'],
             77: ['metre'],
             105: ['metre', 'voltage'],
             106: ['metre', 'voltage'],
             107: ['metre', 'voltage'],
             109: ['metre', 'voltage'],
             112: ['metre'],
             113: ['voltage'],
             40: ['mm']})

In [270]:
# Create dictionary with findings of known_data_properties_in_line
# The dictionary key is the index number of the line

from collections import defaultdict

dictionary_with_known_data_properties_in_a_sentence = defaultdict(int)

for (word_we_know, index_of_lines_with_a_match) in known_data_properties_in_line:
    
    for index_of_line in index_of_lines_with_a_match:
        
        if index_of_line not in dictionary_with_known_data_properties_in_a_sentence:
            
            print('First detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add to dictionary
            dictionary_with_known_data_properties_in_a_sentence[index_of_line] = [word_we_know]
            
        else: # a detection in sentence is already added
            
            print('New detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add if it is a new word
            if word_we_know not in dictionary_with_known_data_properties_in_a_sentence[index_of_line]:
                dictionary_with_known_data_properties_in_a_sentence[index_of_line].append(word_we_know)
                
dictionary_with_known_data_properties_in_a_sentence

First detection in sentence 35 - length
First detection in sentence 64 - length
First detection in sentence 70 - length
First detection in sentence 76 - length
First detection in sentence 77 - length
First detection in sentence 105 - length
First detection in sentence 106 - length
First detection in sentence 107 - length
First detection in sentence 109 - length
First detection in sentence 112 - length
First detection in sentence 48 - closer
First detection in sentence 52 - closer
First detection in sentence 38 - least
First detection in sentence 40 - least
First detection in sentence 41 - least
First detection in sentence 42 - least
First detection in sentence 44 - least
First detection in sentence 45 - least
First detection in sentence 46 - least
First detection in sentence 49 - least
New detection in sentence 77 - least
New detection in sentence 77 - least
New detection in sentence 77 - least
First detection in sentence 78 - least
New detection in sentence 78 - least
First detection 

defaultdict(int,
            {35: ['length'],
             64: ['length'],
             70: ['length'],
             76: ['length'],
             77: ['length', 'least'],
             105: ['length'],
             106: ['length'],
             107: ['length'],
             109: ['length'],
             112: ['length', 'least'],
             48: ['closer'],
             52: ['closer'],
             38: ['least'],
             40: ['least'],
             41: ['least'],
             42: ['least'],
             44: ['least'],
             45: ['least'],
             46: ['least'],
             49: ['least'],
             78: ['least'],
             79: ['least'],
             83: ['least'],
             88: ['least'],
             89: ['least']})

In [271]:
# Merge dictionary_with_known_data_types_in_a_sentence and dictionary_with_known_data_properties_in_a_sentence

dictionary_with_known_data_types_and_properties_in_a_sentence = defaultdict(int)

# Load data type line numbers
for line_number in dictionary_with_known_data_types_in_a_sentence.keys():
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = ([],[])

# Load data property line numbers
for line_number in dictionary_with_known_data_properties_in_a_sentence.keys():
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = ([],[])

# Load data
for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_in_a_sentence[line_number]
    data_properties = dictionary_with_known_data_properties_in_a_sentence[line_number]    
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = (data_types,data_properties)

# Result comment: easy to see what we have.
# A human must look at the sentences with 0 values
    
dictionary_with_known_data_types_and_properties_in_a_sentence

defaultdict(int,
            {38: (['cm'], ['least']),
             42: (['cm'], ['least']),
             44: (['cm'], ['least']),
             45: (['cm'], ['least']),
             48: (['cm'], ['closer']),
             49: (['cm'], ['least']),
             52: (['cm'], ['closer']),
             89: (['cm'], ['least']),
             35: (['metre'], ['length']),
             41: (['metre'], ['least']),
             46: (['metre'], ['least']),
             64: (['metre'], ['length']),
             70: (['metre'], ['length']),
             76: (['metre'], ['length']),
             77: (['metre'], ['length', 'least']),
             105: (['metre', 'voltage'], ['length']),
             106: (['metre', 'voltage'], ['length']),
             107: (['metre', 'voltage'], ['length']),
             109: (['metre', 'voltage'], ['length']),
             112: (['metre'], ['length', 'least']),
             113: (['voltage'], 0),
             40: (['mm'], ['least']),
             78: (0, ['least']),
 

In [272]:
# Focus on sentences with findings on both type AND property

dictionary_with_match_on_type_and_property = defaultdict(int)
dictionary_with_match_on_type_or_property = defaultdict(int)

for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][0]
    data_properties = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][1]
    
    if data_types == 0 or data_properties == 0:
        dictionary_with_match_on_type_or_property[line_number] = (data_types,data_properties)
    else:
        dictionary_with_match_on_type_and_property[line_number] = (data_types,data_properties)

dictionary_with_match_on_type_and_property
#dictionary_with_match_on_type_or_property

defaultdict(int,
            {38: (['cm'], ['least']),
             42: (['cm'], ['least']),
             44: (['cm'], ['least']),
             45: (['cm'], ['least']),
             48: (['cm'], ['closer']),
             49: (['cm'], ['least']),
             52: (['cm'], ['closer']),
             89: (['cm'], ['least']),
             35: (['metre'], ['length']),
             41: (['metre'], ['least']),
             46: (['metre'], ['least']),
             64: (['metre'], ['length']),
             70: (['metre'], ['length']),
             76: (['metre'], ['length']),
             77: (['metre'], ['length', 'least']),
             105: (['metre', 'voltage'], ['length']),
             106: (['metre', 'voltage'], ['length']),
             107: (['metre', 'voltage'], ['length']),
             109: (['metre', 'voltage'], ['length']),
             112: (['metre'], ['length', 'least']),
             40: (['mm'], ['least'])})

In [273]:
#
# Here we have found sentences with something, but type OR property is missing.
# This is a good place for a human to look for sentences with unknown words.
#

dictionary_with_match_on_type_and_property = defaultdict(int)
dictionary_with_match_on_type_or_property = defaultdict(int)

for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][0]
    data_properties = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][1]
    
    if data_types == 0 or data_properties == 0:
        dictionary_with_match_on_type_or_property[line_number] = (data_types,data_properties)
    else:
        dictionary_with_match_on_type_and_property[line_number] = (data_types,data_properties)

#dictionary_with_match_on_type_and_property
dictionary_with_match_on_type_or_property

defaultdict(int,
            {113: (['voltage'], 0),
             78: (0, ['least']),
             79: (0, ['least']),
             83: (0, ['least']),
             88: (0, ['least'])})

# Identify known and unknown things in sentences

In [274]:
#
# Find known words in line
#
# Now we have identified lines with known types and properties.
# Next up, can we identify the thing we are talking about?
#
# Idea: Remove the identified words from the sentences before continuing to searching in them

selected_line_in_a_list_of_words = []

for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:
    if line_index_number == 105:
        selected_line_in_a_list_of_words.append((word_index_number,line_index_number,word_in_text,pos))

line_with_findings_removed = [(word_index_number,line_index_number,word_in_text,pos) for (word_index_number,line_index_number,word_in_text,pos) in selected_line_in_a_list_of_words if word_in_text not in list_of_known_data_types]
line_with_findings_removed

[(1047, 105, 'electrical', 'JJ'),
 (1048, 105, 'installation', 'NNS'),
 (1049, 105, 'vessel', 'NNP'),
 (1051, 105, 'overall', 'JJ'),
 (1052, 105, 'length', 'NN'),
 (1053, 105, 'construct', 'VBD'),
 (1054, 105, 'electrical', 'JJ'),
 (1055, 105, 'installation', 'NNS'),
 (1058, 105, 'satisfy', 'VB'),
 (1059, 105, 'requirement', 'NNS'),
 (1060, 105, 'nordic', 'NNP'),
 (1061, 105, 'boat', 'NNP'),
 (1062, 105, 'standard', 'NNP'),
 (1063, 105, 'electrical', 'JJ'),
 (1064, 105, 'installation', 'NNS')]

In [275]:
#
# Find sentences where we are going to look for something (the subject in the RDf triple)
#
# Now we have identified lines with known types and properties.
# Next up, can we identify the thing we are talking about?
#
# Idea: Remove the identified words from the sentences before continuing to searching in them

sentences_without_data_type_and_data_property = []

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:
        
        # only add words from sentence that is not data_type or data_property
        data_types = dictionary_with_match_on_type_and_property[line_number][0]
        data_properties = dictionary_with_match_on_type_and_property[line_number][1]

        if line_index_number == line_number and word_in_text not in data_types and word_in_text not in data_properties:
            sentences_without_data_type_and_data_property.append((word_index_number,line_index_number,word_in_text,pos))

sentences_without_data_type_and_data_property

[(336, 38, 'stove', 'NNS'),
 (337, 38, 'wooden', 'JJ'),
 (338, 38, 'floor', 'NN'),
 (339, 38, 'combustible', 'JJ'),
 (340, 38, 'floor', 'NN'),
 (341, 38, 'cover', 'VBG'),
 (342, 38, 'protect', 'VBN'),
 (343, 38, 'steel', 'NN'),
 (344, 38, 'aluminium', 'NN'),
 (345, 38, 'sheet', 'NNS'),
 (346, 38, 'cover', 'VBG'),
 (347, 38, 'whole', 'JJ'),
 (348, 38, 'floor', 'NN'),
 (349, 38, 'area', 'NN'),
 (350, 38, 'fireplace', 'NN'),
 (351, 38, 'front', 'JJ'),
 (352, 38, 'stove', 'NN'),
 (355, 38, 'fire', 'NN'),
 (356, 38, 'door', 'NN'),
 (387, 42, 'saucer', 'NNP'),
 (388, 42, 'form', 'VBD'),
 (389, 42, 'rockwool', 'JJ'),
 (390, 42, 'sheet', 'NNS'),
 (393, 42, 'thickness', 'NN'),
 (394, 42, 'provide', 'VB'),
 (395, 42, 'smoke', 'NN'),
 (396, 42, 'uptake', 'JJ'),
 (397, 42, 'combustible', 'JJ'),
 (398, 42, 'material', 'NN'),
 (409, 44, 'alternatively', 'RB'),
 (410, 44, 'pipe', 'JJ'),
 (411, 44, 'diameter', 'NN'),
 (414, 44, 'large', 'JJR'),
 (415, 44, 'smoke', 'NN'),
 (416, 44, 'uptake', 'VBP'),
 

## Find single word suggestions

In [276]:
test = [(1057, 105, 'electrical', 'JJ'),
 (1058, 105, 'installation', 'NNS'),
 (1059, 105, 'vessel', 'NNP'),
 (1061, 105, 'overall', 'JJ'),
 (1063, 105, 'construct', 'VBD'),
 (1065, 105, 'electrical', 'JJ'),
 (1066, 105, 'installation', 'NNS'),
 (1069, 105, 'satisfy', 'VB'),
 (1070, 105, 'requirement', 'NNS'),
 (1071, 105, 'nordic', 'NNP'),
 (1072, 105, 'boat', 'NNP'),
 (1073, 105, 'standard', 'NNP'),
 (1074, 105, 'electrical', 'JJ'),
 (1075, 105, 'installation', 'NNS')]

# her ser ein eit mønster 1071 + 1072 + 1073 == nordic boat standard

# TODO: create CSV with suggestions of things of new things

# TODO: suggest things we know, that we find, list_of_known_things = ['vessel', 'emergency light', 'pipe system', 'pipe', 'fire extinguisher']

for (word_index_number,line_index_number,word_in_text,pos) in test:
    if pos.startswith('NNP'):
        print((word_index_number,line_index_number,word_in_text,pos))

(1059, 105, 'vessel', 'NNP')
(1071, 105, 'nordic', 'NNP')
(1072, 105, 'boat', 'NNP')
(1073, 105, 'standard', 'NNP')


In [277]:
# print the nouns
for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            print((word_index_number,line_index_number,word_in_text,pos))

(336, 38, 'stove', 'NNS')
(338, 38, 'floor', 'NN')
(340, 38, 'floor', 'NN')
(343, 38, 'steel', 'NN')
(344, 38, 'aluminium', 'NN')
(345, 38, 'sheet', 'NNS')
(348, 38, 'floor', 'NN')
(349, 38, 'area', 'NN')
(350, 38, 'fireplace', 'NN')
(352, 38, 'stove', 'NN')
(355, 38, 'fire', 'NN')
(356, 38, 'door', 'NN')
(387, 42, 'saucer', 'NNP')
(390, 42, 'sheet', 'NNS')
(393, 42, 'thickness', 'NN')
(395, 42, 'smoke', 'NN')
(398, 42, 'material', 'NN')
(411, 44, 'diameter', 'NN')
(415, 44, 'smoke', 'NN')
(417, 44, 'pipe', 'NN')
(420, 44, 'order', 'NN')
(423, 44, 'air', 'NN')
(424, 44, 'circulation', 'NN')
(425, 44, 'pipe', 'NNS')
(426, 45, 'insulation', 'NNP')
(427, 45, 'saucer', 'NN')
(431, 45, 'wool', 'NN')
(432, 45, 'twine', 'NN')
(435, 45, 'thickness', 'NN')
(437, 45, 'outer', 'NN')
(438, 45, 'pipe', 'NN')
(440, 45, 'material', 'NN')
(457, 48, 'smoke', 'NNP')
(458, 48, 'uptake', 'NNS')
(463, 48, 'bulkhead', 'NNS')
(464, 49, 'smoke', 'NN')
(466, 49, 'pipe', 'NN')
(468, 49, 'saucer', 'NN')
(472, 49

In [278]:
# TODO: compare to list of known things and print out to csv
# TODO: do this for 1, 2, 3 and 4 word combinations
# TODO: check words for repeating words and suggest thouse
# TODO: create fragments of a sentence, and create bi and thrigrams of thouse fragments

In [279]:
# Look for words that we know

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            if word_in_text in list_of_known_things:
                print((word_index_number,line_index_number,word_in_text,pos))

(417, 44, 'pipe', 'NN')
(425, 44, 'pipe', 'NNS')
(438, 45, 'pipe', 'NN')
(466, 49, 'pipe', 'NN')
(498, 52, 'pipe', 'NNS')
(318, 35, 'vessel', 'NNS')
(441, 46, 'pipe', 'NN')
(630, 64, 'vessel', 'NNP')
(674, 70, 'vessel', 'NNP')
(742, 76, 'vessel', 'NNP')
(763, 77, 'vessel', 'NNP')
(1049, 105, 'vessel', 'NNP')
(1065, 106, 'vessel', 'NNS')
(1079, 107, 'vessel', 'NNS')
(1097, 109, 'vessel', 'NNS')
(1124, 112, 'vessel', 'NNS')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')


Comment on the result:
Are we only alking about "vessel" here?
It is the only recognised word in our test vocabulary / ontology

(740, 76, 'vessel', 'NNP')
(1047, 105, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')

-- Update: Added 'least' to list_of_known_data_properties. New result:

(416, 44, 'pipe', 'NN')
(424, 44, 'pipe', 'NNS')
(437, 45, 'pipe', 'NN')
(464, 49, 'pipe', 'NN')
(740, 76, 'vessel', 'NNP')
(1047, 105, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')

Here we can see that 'pipe' is identified as something we know.


 -- Update: Added custom lemmatizer on 'metres' to 'metre'
 
 (416, 44, 'pipe', 'NN')
(424, 44, 'pipe', 'NNS')
(437, 45, 'pipe', 'NN')
(464, 49, 'pipe', 'NN')
(318, 35, 'vessel', 'NNS')
(628, 64, 'vessel', 'NNP')
(672, 70, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')
(1122, 112, 'vessel', 'NNS')
(1047, 105, 'vessel', 'NNP')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')

In [280]:
# Look for words that we know

test = []

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            if word_in_text in list_of_known_things:
                data_types = dictionary_with_match_on_type_and_property[line_number][0]
                data_properties = dictionary_with_match_on_type_and_property[line_number][1]
                
                for data_type in data_types:
                    
                    for data_property in data_properties:
                        
                        test.append(word_in_text + "," + data_property + "," + data_type)

test

['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,closer,cm',
 'vessel,length,metre',
 'pipe,least,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,least,metre',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,least,metre',
 'pipe,least,mm',
 'pipe,least,mm']

Comment on result

['vessel,length,metres',
 'vessel,length,metres',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v']
 
 We are only talking about one thing here, and we have a lot of duplicates form the different sentences.
 
-- Update: Added 'least' to list_of_known_data_properties. New result:
 
 ['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'vessel,length,metres',
 'vessel,length,metres',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'pipe,least,mm',
 'pipe,least,mm']
 
 -- Update: Added custom lemmatizer on 'metres' to 'metre'
 
 ['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,least,metre',
 'vessel,length,v',
 'pipe,least,mm',
 'pipe,least,mm']
 
 TODO: Length and voltage do not belong together

In [281]:
# Removing duplicates from result and make CSV output ready

output_without_duplicates = list(dict.fromkeys(test))
lines_for_csv_file = []

for line in output_without_duplicates:
    line_split = line.split(",")
    lines_for_csv_file.append((line_split[0],line_split[1],line_split[2]))

lines_for_csv_file

[('pipe', 'least', 'cm'),
 ('pipe', 'closer', 'cm'),
 ('vessel', 'length', 'metre'),
 ('pipe', 'least', 'metre'),
 ('vessel', 'least', 'metre'),
 ('vessel', 'length', 'voltage'),
 ('pipe', 'least', 'mm')]

In [282]:
import csv

with open("output - Identify data type properties - Single words.csv", 'w') as f:
    write = csv.writer(f, delimiter=',', lineterminator='\n')
    write.writerows(lines_for_csv_file)